robolectric/src/main/java/org/robolectric/Robolectric.java

```diff
@@ -378,4 +385,75 @@ public static Scheduler getBackgroundThreadScheduler() {
   public static void flushBackgroundThreadScheduler() {
     getBackgroundThreadScheduler().advanceToLastPostedRunnable();
   }
+
+  @SuppressWarnings({"NewApi", "unchecked"})
+  private static <T extends Service> T instantiateService(Class<T> serviceClass, Intent intent) {
+    if (RuntimeEnvironment.getApiLevel() >= P) {
+      final LoadedApk loadedApk = getLoadedApk();
+      AppComponentFactory factory = getAppComponentFactory(loadedApk);
+      if (factory != null) {
+        try {
+          return (T)
+              factory.instantiateService(
+                  loadedApk.getClassLoader(), serviceClass.getName(), intent);
+        } catch (ReflectiveOperationException e) {
+          Logger.debug("Failed to instantiate Service using AppComponentFactory", e);
+        }
+      }
+    }
+    return ReflectionHelpers.callConstructor(serviceClass);
+  }
+
+  @SuppressWarnings({"NewApi", "unchecked"})
+  private static <T extends ContentProvider> T instantiateContentProvider(Class<T> providerClass) {
+    if (RuntimeEnvironment.getApiLevel() >= P) {
+      final LoadedApk loadedApk = getLoadedApk();
+      AppComponentFactory factory = getAppComponentFactory(loadedApk);
+      if (factory != null) {
+        try {
+          return (T)
+              factory.instantiateProvider(loadedApk.getClassLoader(), providerClass.getName());
```

Comment:
I don't think that `instantiateProvider` will be the thing throwing the error.

Instead of catching ReflectiveOperationException, could there be an instanceof check?